pennylane/templates/subroutines/qmc.py

```diff
@@ -326,32 +325,47 @@ def circuit():
         >>> (1 - np.cos(np.pi * phase_estimated)) / 2
         0.4327096457464369
     """
-    if isinstance(probs, np.ndarray) and probs.ndim != 1:
-        raise ValueError("The probability distribution must be specified as a flat array")
-
-    dim_p = len(probs)
-    num_target_wires = float(np.log2(2 * dim_p))
-
-    if not num_target_wires.is_integer():
-        raise ValueError("The probability distribution must have a length that is a power of two")
-
-    num_target_wires = int(num_target_wires)
-
-    target_wires = Wires(target_wires)
-    estimation_wires = Wires(estimation_wires)
-
-    if num_target_wires != len(target_wires):
-        raise ValueError(
-            f"The probability distribution of dimension {dim_p} requires"
-            f" {num_target_wires} target wires"
-        )
-
-    A = probs_to_unitary(probs)
-    R = func_to_unitary(func, dim_p)
-    Q = make_Q(A, R)
-
-    qml.QubitUnitary(A, wires=target_wires[:-1])
-    qml.QubitUnitary(R, wires=target_wires)
-    qml.templates.QuantumPhaseEstimation(
-        Q, target_wires=target_wires, estimation_wires=estimation_wires
-    )
+    num_params = 3
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, probs, func, target_wires, estimation_wires, do_queue=True):
```

Comment:
I thought there was a `_pre_process()` method that should be used to perform checks on inputs? i.e., these checks shouldn't happen inside `__init__`?

Reply:
`_pre_process()` isn't something defined in the `Operation` base class - it's more something defined for templates. We didn't define a `_pre_process()` function for this `QuantumMonteCarlo` template though. I'm not sure if it's considered a requirement to have `_pre_process()` in the `Operation` version?

For `QuantumMonteCarlo`, I don't really see a problem of doing input validation in `__init__`, and there is a precedent for doing so in some of the core ops (e.g., `PauliRot`).

But I don't mind moving to a `_pre_process()` method if we want that for all templates that become ops.

Reply:
`_pre_process` is definitely not a hard requirement as far as I am aware, but exists partly because at one point, all templates were refactored by @mariaschuld and so they follow the same design pattern.

@mariaschuld, would you prefer all templates stick to a `_pre_process()` method (we could even make it part of the API?), or just leave the implementation details up to the developer?

Reply:
Not at all, I consider `_preprocess` just as a convenient way to outsource all sorts of input checks to a separate function, but it could be done in the init method - especially if its relatively straight forward.